cEP-0002.md

```diff
@@ -0,0 +1,402 @@
+
+        # May raise RuntimeError so bear doesn't get executed on invalid params
+        self.kwargs = get_kwargs_for_function(self.run, section)
+
+    def execute_task(self, task, **kwargs):
+        """
+        This method is responsible for getting results from the
+        analysis routines and populating the result queue.
+        """
+        return list(self.analyze(task, **self.kwargs))
+
+    def analyze(self,
+                file_proxy,
+                *args,
+                dependency_results=None,
+                **kwargs):
+        """
+        Handles the given file.
+
+        :param file_proxy: Object containing filename and contents
+        :return:           A list of Result
+        """
+        raise NotImplementedError("This function has to be implemented for a "
+                                  "runnable bear.")
+
+    def generate_tasks(self):
+        """
+        This method is responsible for providing the files.
+        """
+        return self.file_set
+```
+
+A test bear:
+
+```python
+class TestBear(FileWiseBear):
+    def analyze(self, file_proxy):
+        """
+        Handles the given file.
+
+        :param file_proxy: Object containing filename and contents
+        :return:           A list of Result
+        """
+        for i, line in enumerate(file):
+            # This will contain the analysis routine
+            yield "Line has a problem" , file_proxy.name, i
+```
+
+A rough idea of how the new bear will be called:
+
+```python
+import asyncio
+import concurrent.futures
+import logging
+import sys
+from coalib.collecting.Collectors import collect_files
+from Fileproxy import FileProxy
+from TestBear import TestBear
+from AnotherBear import AnotherBear
+from Dependencies import Dependencies
+from DependentBear import DependentBear
+from AnotherDependentBear import AnotherDependentBear
+import functools
+import multiprocessing
+
+
+def get_cpu_count():
+    try:
+        return multiprocessing.cpu_count()
+    # cpu_count is not implemented for some CPU architectures/OSes
+    except NotImplementedError:  # pragma: no cover
+        return 2
+
+
+def schedule_bears(bears,
+                   dep_resolver,
+                   event_loop,
+                   blocking_tasks,
+                   executor):
+    """
+    Schedules the tasks of bears to the process pool and run the event loop to
+    complete those tasks.
+    :param bears:          A list of bears to be scheduled onto the process
+                           pool.
+    :param dep_resolver:   The object that keeps track of dependencies.
+    :param event_loop:     The asyncio event loop.
+    :param blocking_tasks: Dictionary that keeps track of the remaining tasks
+                           of each bear.
+    :param executor:       The process pool to which the bear tasks are
+                           scheduled.
+    """
+    for bear in bears:
+        if bear not in dep_resolver.dependency_dict:
+            print(bear)
```

Comment:
can you remove debug prints on your own? I would recommend to self review this PR at least once btw.